core/rfb.js

```diff
@@ -482,60 +456,66 @@
                 this._sock.off('close');  // make sure we don't get a double event
             }
 
+            this._onUpdateState(this, state, oldstate);
             switch (state) {
+                case 'disconnected':
+                    this._cleanupSocket(); // Make sure we are closed.
+
+                    if (typeof(errorMsg) !== 'undefined') {
+                        this._onDisconnected(this, errorMsg);
+                    } else {
+                        this._onDisconnected(this); // No msg means clean disconnect
+                    }
+                    break;
+
                 case 'connected':
-                    if (oldstate === 'disconnected' || oldstate === 'failed') {
-                        Util.Error("Invalid transition from 'disconnected' or 'failed' to 'connected'");
+                    if (oldstate === 'disconnected') {
+                        Util.Error("Invalid transition from 'disconnected' to 'connected'");
                     }
                     break;
 
                 case 'connecting':
+                    // A previous connect may asynchronously cause a connection
+                    // so make sure we are still closed.
+                    this._cleanupSocket();
```

Comment:
It seems like the browsers have some issues in displaying active event listeners properly in their development tools. All listeners that are added on the Mouse and Keyboard objects (devices.js) are properly removed on cleanup. Events are only added in the grab functions and are all removed in the ungrab functions (which are called on cleanup).

    $ git grep -n -e "addEventListen" -e "removeEventListen" core/input/devices.js
    core/input/devices.js:125:            c.addEventListener('keydown',     this._eventHandlers.keydown);
    core/input/devices.js:126:            c.addEventListener('keyup', this._eventHandlers.keyup);
    core/input/devices.js:127:            c.addEventListener('keypress', this._eventHandlers.keypress);
    core/input/devices.js:130:            window.addEventListener('blur', this._eventHandlers.blur);
    core/input/devices.js:139:            c.removeEventListener('keydown', this._eventHandlers.keydown);
    core/input/devices.js:140:            c.removeEventListener('keyup', this._eventHandlers.keyup);
    core/input/devices.js:141:            c.removeEventListener('keypress', this._eventHandlers.keypress);
    core/input/devices.js:142:            window.removeEventListener('blur', this._eventHandlers.blur);
    core/input/devices.js:350:                c.addEventListener('touchstart', this._eventHandlers.mousedown);
    core/input/devices.js:351:                window.addEventListener('touchend', this._eventHandlers.mouseup);
    core/input/devices.js:352:                c.addEventListener('touchend', this._eventHandlers.mouseup);
    core/input/devices.js:353:                c.addEventListener('touchmove', this._eventHandlers.mousemove);
    core/input/devices.js:355:                c.addEventListener('mousedown', this._eventHandlers.mousedown);
    core/input/devices.js:356:                window.addEventListener('mouseup', this._eventHandlers.mouseup);
    core/input/devices.js:357:                c.addEventListener('mouseup', this._eventHandlers.mouseup);
    core/input/devices.js:358:                c.addEventListener('mousemove', this._eventHandlers.mousemove);
    core/input/devices.js:359:                c.addEventListener((Util.Engine.gecko) ? 'DOMMouseScroll' : 'mousewheel',
    core/input/devices.js:364:            document.addEventListener('click', this._eventHandlers.mousedisable);
    core/input/devices.js:365:            document.body.addEventListener('contextmenu', this._eventHandlers.mousedisable);
    core/input/devices.js:372:                c.removeEventListener('touchstart', this._eventHandlers.mousedown);
    core/input/devices.js:373:                window.removeEventListener('touchend', this._eventHandlers.mouseup);
    core/input/devices.js:374:                c.removeEventListener('touchend', this._eventHandlers.mouseup);
    core/input/devices.js:375:                c.removeEventListener('touchmove', this._eventHandlers.mousemove);
    core/input/devices.js:377:                c.removeEventListener('mousedown', this._eventHandlers.mousedown);
    core/input/devices.js:378:                window.removeEventListener('mouseup', this._eventHandlers.mouseup);
    core/input/devices.js:379:                c.removeEventListener('mouseup', this._eventHandlers.mouseup);
    core/input/devices.js:380:                c.removeEventListener('mousemove', this._eventHandlers.mousemove);
    core/input/devices.js:381:                c.removeEventListener((Util.Engine.gecko) ? 'DOMMouseScroll' : 'mousewheel',
    core/input/devices.js:386:            document.removeEventListener('click', this._eventHandlers.mousedisable);
    core/input/devices.js:387:            document.body.removeEventListener('contextmenu', this._eventHandlers.mousedisable);